rfcs/2021-10-29-9465-splunk-hec-indexer-acknowledgements.md

```diff
@@ -0,0 +1,346 @@
+    the data, we will use the existing `BatchNotifier`/`BatchReceiver` system
+    from vector’s overall end-to-end-acknowledgement infrastructure.
+
+    ```{rust}
+    async fn handle_request(acknowledgements: bool, events: Vec<Events>, out: Pipeline, ack_id: usize...) {
+        let receiver = acknowledgements.then(|| {
+                let (batch, receiver) = BatchNotifier::new_with_receiver();
+                for event in &mut events {
+                    event.add_batch_notifier(Arc::clone(&batch));
+                }
+                receiver
+            });
+
+            out.send_all(&mut futures::stream::iter(events).map(Ok))
+                ...
+                .and_then(|_| handle_batch_status(ack_id, receiver))
+                .await
+    }
+    ...
+    async fn handle_batch_status(ack_id, receiver) {
+        match receiver.await {
+            BatchStatus::Delivered => // update ackId -> true,
+            BatchStatus::Errored => // leave ackId -> false,
+            BatchStatus::Failed => // leave ackId -> false,
+        }
+    }
+    ```
+
+  * When the data in the request is successfully delivered, we mark
+    `ack_ids_ack_status[currently_available_ack_id] = 1`.
+* Querying ackId status
+  * We add a new `/services/collector/ack` route
+  * `ackId`’s from incoming requests are used to index into
+    `ack_ids_ack_status`. We return true/false depending on the value in the
+    bitvec.
+  * If we return true for an ackId, we mark `ack_ids_in_use[ack_id] = 0` and
+    `ack_ids_ack_status[ack_id] = 0`. The `ackId` can then be reused.
+
+## Proposal: `splunk_hec` Sinks Indexer Acknowledgement
+
+### User Experience
+
+The `splunk_hec` sinks will automatically integrate with indexer
+acknowledgements if the user has enabled it in their Splunk instance (if
+`ackId`’s are present in HEC responses). If indexer acknowledgement is disabled,
+`splunk_hec` sinks will continue to finalize events based on the HEC response
+status code.
+
+### Implementation
+
+* [Splunk recommendations for client integration with indexer acknowledgement](https://docs.splunk.com/Documentation/Splunk/8.2.3/Data/AboutHECIDXAck#Indexer_acknowledgment_client_behavior)
+
+We generate a single `guid` to use as a channel ID and include this in HEC
+requests. After submitting events to HEC, we will parse the HTTP response for an
+`ackId`. If no `ackId` is found, we rely on the current behavior of setting
+`EventStatus` based solely on the HTTP status code.
+
+If an `ackId` is found, we store it in a `Map<u64, bool>` shared with a
+background tokio task which, for all pending `ackId`’s, will query
+`/services/collector/ack`. This background task will query at an interval of 10
+seconds (as recommended by Splunk) and update the shared structure based on the
+results. Meanwhile, back in the HEC response handler, we’ll check the shared
+structure for `true` at 10 second intervals up to a maximum time limit (5
+minutes as recommended by Splunk).
```

Comment:
I'm wondering if we'd like to leave this configurable (with limits?) to manage resource usage at Vector (similar to the max acks option)

Reply:
Added in 2fc7df5